arcgis-ios-sdk-samples/Maps/Generate offline map (basemap by reference)/GenerateOfflineMapBasemapByReferenceViewController.swift

```diff
@@ -0,0 +1,225 @@
+
+/// A view controller that manages the interface of the Generate Offline Map
+/// (Basemap by Reference) sample.
+class GenerateOfflineMapBasemapByReferenceViewController: UIViewController {
+    /// The map view managed by the view controller.
+    @IBOutlet weak var mapView: AGSMapView! {
+        didSet {
+            mapView.map = makeMap()
+            // Set inset of overlay.
+            let padding: CGFloat = 30
+            mapView.layoutMargins = UIEdgeInsets(top: padding, left: padding, bottom: padding, right: padding)
+        }
+    }
+    /// The view used to show the geographic area for which the map data should
+    /// be taken offline.
+    @IBOutlet weak var areaOfInterestView: UIView!
+    /// The button item that starts generating an offline map.
+    @IBOutlet weak var generateButtonItem: UIBarButtonItem!
+    
+    /// Creates a map.
+    ///
+    /// - Returns: A new `AGSMap` object.
+    func makeMap() -> AGSMap {
+        let portal = AGSPortal.arcGISOnline(withLoginRequired: false)
+        let portalItem = AGSPortalItem(portal: portal, itemID: "acc027394bc84c2fb04d1ed317aac674")
+        let map = AGSMap(item: portalItem)
+        map.load { [weak self] _ in self?.mapDidLoad() }
+        return map
+    }
+    
+    /// Called in response to the map load operation completing.
+    func mapDidLoad() {
+        guard let map = mapView.map else { return }
+        if let error = map.loadError {
+            presentAlert(error: error)
+        } else {
+            generateButtonItem.isEnabled = true
+        }
+    }
+    
+    /// Called in response to the Generate Offline Map button item being tapped.
+    @IBAction func generateOfflineMap() {
+        generateButtonItem.isEnabled = false
+        mapView.isUserInteractionEnabled = false
+        let offlineMapTask = AGSOfflineMapTask(onlineMap: mapView.map!)
+        offlineMapTask.defaultGenerateOfflineMapParameters(withAreaOfInterest: areaOfInterest) { [weak self] (parameters, error) in
+            if let parameters = parameters {
+                self?.offlineMapTask(offlineMapTask, didCreate: parameters)
+            } else if let error = error {
+                self?.offlineMapTask(offlineMapTask, didFailToCreateParametersWith: error)
+            }
+        }
+    }
+    
+    func offlineMapTask(_ offlineMapTask: AGSOfflineMapTask, didCreate parameters: AGSGenerateOfflineMapParameters) {
+        let filename = parameters.referenceBasemapFilename as NSString
+        let name = filename.deletingPathExtension
+        let `extension` = filename.pathExtension
+        if let url = Bundle.main.url(forResource: name, withExtension: `extension`) {
+            let message: String = {
+                let displayName = Bundle.main.object(forInfoDictionaryKey: "CFBundleDisplayName") as! String
+                return String(format: "“%@” has a local version of the basemap used by the map. Would you like to use the local or the online basemap?", displayName)
```

Comment:
Good question. String interpolation would certainly work here, but I'm wary of setting a bad example. Constructing user-facing strings using string interpolation only works in non-localized environments, whereas using a format string works equally well in both localized and non-localized environments. Same with using `Bundle.object(forInfoDictionaryKey:)` in the previous line ("Use of this method is preferred over other access methods because it returns the localized value of a key when one is available.").

Does that make sense?